src/packet/public_key_encrypted_session_key.rs

```diff
@@ -13,95 +13,252 @@ use crate::crypto::sym::SymmetricKeyAlgorithm;
-use crate::types::{mpi, KeyId, Mpi, PublicKeyTrait, Tag, Version};
+use crate::types::{
+    mpi, EskType, Fingerprint, KeyId, KeyVersion, PkeskBytes, PublicKeyTrait, PublicParams, Tag,
+    Version,
+};
 
-/// Public Key Encrypted Session Key Packet
-/// <https://tools.ietf.org/html/rfc4880.html#section-5.1>
+/// Public Key Encrypted Session Key Packet (PKESK)
+/// <https://www.rfc-editor.org/rfc/rfc9580.html#name-public-key-encrypted-sessio>
+///
+/// A PKESK contains an encrypted session key that has been encrypted to a specific public key.
+/// PKESK are used in combination with a symmetric encryption container:
+///
+/// - V3 PKESK are used in combination with [version 1 Symmetrically Encrypted and Integrity
+///   Protected Data Packets](https://www.rfc-editor.org/rfc/rfc9580.html#name-version-1-symmetrically-enc).
+/// - V6 PKESK are used in combination with [version 2 Symmetrically Encrypted and Integrity
+///   Protected Data Packets](https://www.rfc-editor.org/rfc/rfc9580.html#name-version-2-symmetrically-enc).
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct PublicKeyEncryptedSessionKey {
-    packet_version: Version,
-    version: u8,
-    id: KeyId,
-    algorithm: PublicKeyAlgorithm,
-    mpis: Vec<Mpi>,
+pub enum PublicKeyEncryptedSessionKey {
+    V3 {
+        packet_version: Version,
+        id: KeyId,
+        pk_algo: PublicKeyAlgorithm,
+        values: PkeskBytes,
+    },
+
+    V6 {
+        packet_version: Version,
+        fingerprint: Option<Fingerprint>,
+        pk_algo: PublicKeyAlgorithm,
+        values: PkeskBytes,
+    },
+
+    Other {
+        packet_version: Version,
+        version: u8,
+    },
 }
 
 impl PublicKeyEncryptedSessionKey {
     /// Parses a `PublicKeyEncryptedSessionKey` packet from the given slice.
     pub fn from_slice(version: Version, input: &[u8]) -> Result<Self> {
         let (_, pk) = parse(version)(input)?;
 
-        if pk.version != 3 {
-            unsupported_err!("unsupported PKESK version {}", pk.version);
-        }
-
         Ok(pk)
     }
 
-    /// Encrypts the given session key to the passed in public key.
-    pub fn from_session_key<R: CryptoRng + Rng>(
+    /// Encrypts the given session key to `pkey` as a v3 pkesk.
+    pub fn from_session_key_v3<R: CryptoRng + Rng>(
         rng: R,
         session_key: &[u8],
         alg: SymmetricKeyAlgorithm,
         pkey: &impl PublicKeyTrait,
     ) -> Result<Self> {
         // the session key is prefixed with symmetric key algorithm
         let len = session_key.len();
-        let mut data = vec![0u8; len + 3];
+        let mut data = vec![0u8; len + 1];
         data[0] = u8::from(alg);
         data[1..=len].copy_from_slice(session_key);
 
-        // and appended a checksum
-        BigEndian::write_u16(
-            &mut data[len + 1..],
-            checksum::calculate_simple(session_key),
-        );
+        // Append a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
```

Comment:
what about this?

Reply:
Looking into this sent me down a surprising rabbit-hole of zeroizing and DRY-ing: 2fb88ac6a3863f9cfd2d9ad78842861aa2c86b36

Please check if all of these changes looks reasonable :sweat_smile: